server.js

```diff
@@ -785,41 +797,97 @@ cache(function (data, match, sendBadge, request) {
+
+  camp.route(new RegExp(fullRegex),
+  cache(function(data, match, sendBadge, request) {
+    // Assumes the final capture group is the extension
+    const format = match.pop();
+    // Discard the first capture group as it's the entire match
+    const params = match.slice(1);
+
+    const badgeData = getBadgeData(config.type, data);
+    config.handler(params, request.asPromise)
+      .then(serviceData => {
+        let text = badgeData.text;
+        if (serviceData.text) {
+          text[1] = serviceData.text;
+        }
+        Object.assign(badgeData, serviceData);
+        badgeData.text = text;
+        sendBadge(format, badgeData);
+      })
+      .catch(error => {
+        console.error(error);
+        const text = badgeData.text;
+        text[1] = 'error';
+        badgeData.text = text;
+        sendBadge(format, badgeData);
+      });
+  }));
+}
+
 // AppVeyor CI integration.
-camp.route(/^\/appveyor\/ci\/([^\/]+\/[^\/]+)(?:\/(.+))?\.(svg|png|gif|jpg|json)$/,
-cache(function(data, match, sendBadge, request) {
-  var repo = match[1];  // eg, `gruntjs/grunt`.
-  var branch = match[2];
-  var format = match[3];
-  var apiUrl = 'https://ci.appveyor.com/api/projects/' + repo;
-  if (branch != null) {
-    apiUrl += '/branch/' + branch;
-  }
-  var badgeData = getBadgeData('build', data);
-  request(apiUrl, { headers: { 'Accept': 'application/json' } }, function(err, res, buffer) {
-    if (err != null) {
-      badgeData.text[1] = 'inaccessible';
-      sendBadge(format, badgeData);
+registerService({
+  name: 'AppVeyor',
+  type: 'build',
+  uri: '/appveyor/ci/([^/]+/[^/]+)(?:/(.+))?',
+  examples: [
+    {uri: '/appveyor/ci/gruntjs/grunt'},
+    {
+      name: 'Branch',
+      params: '/appveyor/ci/gruntjs/grunt/master',
+    },
+  ],
+  handler: async ([repo, branch], request) => {
+    let apiUrl = 'https://ci.appveyor.com/api/projects/' + repo;
+    if (branch != null) {
+      apiUrl += '/branch/' + branch;
     }
+
+    let result;
     try {
-      var data = JSON.parse(buffer);
-      var status = data.build.status;
+      result = await request(apiUrl, {
+        headers: { 'Accept': 'application/json' }
+      });
+    } catch (e) {
+      return {text: 'inaccessible'};
+    }
+
+    try {
+      const data = JSON.parse(result.buffer);
+      const status = data.build.status;
       if (status === 'success') {
-        badgeData.text[1] = 'passing';
-        badgeData.colorscheme = 'brightgreen';
+        return {text: 'passing', colorscheme: 'brightgreen'};
       } else if (status !== 'running' && status !== 'queued') {
-        badgeData.text[1] = 'failing';
-        badgeData.colorscheme = 'red';
+        return {text: 'failing', colorscheme: 'red'};
       } else {
-        badgeData.text[1] = status;
+        return {text: status};
       }
-      sendBadge(format, badgeData);
     } catch(e) {
-      badgeData.text[1] = 'invalid';
-      sendBadge(format, badgeData);
+      return {text: 'invalid'};
```

Comment:
I’ve been reflecting on the way we deal with malformed responses with a mix of careful (e.g. `Array.isArray`) and broad (try/catch) checks. A try/catch in particular makes it hard for service users (or contributors/maintainers) to differentiate between programmer errors and malformed upstream responses. When writing new code, it creates gotchas, because the try block swallows up all sorts of things like `ReferenceError`s.

There’s also a frankensteining of styles: sometimes we have fanatical checks, other times we rely on the try block, and sometimes it's both at the same time.

I wonder if a declarative approach to validation would improve this, and at the same time reduce the number of branches.